File: py/structure/linear/linked_list/DoublyLinkedList.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class DoublyLinkedList:
    def __init__(self):
        self.__head, self.__tail = self.__ListNode(0), self.__ListNode(0)
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    def isEmpty(self) -> bool:
        return self.__head.next == self.__tail

    def add(self, index: int, value: int) ->None:
        current, count = self.__head, 0
        while current != self.__tail and count < index:
            current = current.next
            count += 1

        if current != self.__tail and count <= index:
            node = self.__ListNode(value)
            node.next, node.prev, current.next.prev, current.next = current.next, current, node, node

    def addHead(self, value: int) -> None:
        node = self.__ListNode(value)
        node.next, node.prev, self.__head.next.prev, self.__head.next = self.__head.next, self.__head, node, node

    def addTail(self, value: int) -> None:
        node = self.__ListNode(value)
        node.prev, node.next, self.__tail.prev.next, self.__tail.prev = self.__tail.prev, self.__tail, node, node

    def erase(self, index: int) -> None:
        current, count = self.__head, 0
        while current.next != self.__tail and count < index:
            current = current.next
            count += 1

        if current.next != self.__tail and count <= index:
            current.next.next.prev, current.next = current, current.next.next

    def eraseHead(self) -> None:
        if self.__head.next != self.__tail:
            self.__head.next.next.prev, self.__head.next = self.__head, self.__head.next.next

    def eraseTail(self) -> None:
        if self.__head.next != self.__tail:
            self.__tail.prev.prev.next, self.__tail.prev = self.__tail, self.__tail.prev.prev

    def set(self, index: int, value: int) -> None:
        current, count = self.__head.next, 0
        while current and count < index:
            current = current.next
            count += 1

        if current != self.__tail and count <= index:
            current.value = value

    def get(self, index: int) -> int:
        current, count = self.__head.next, 0
        while current and count < index:
            current = current.next
            count += 1

        if current != self.__tail and count <= index:
            return current.value
        return -sys.maxsize

    def clear(self) -> None:
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    class __ListNode:
        def __init__(self, value: int):
            self.value, self.prev, self.next = value, None, None

File: java/structure/linear/linked_list/DoublyLinkedList.java
package structure.linear.linked_list;

public class DoublyLinkedList {
    private ListNode _head = new ListNode(0);
    private ListNode _tail = new ListNode(0);

    public DoublyLinkedList() {
        _head.next = _tail;
        _tail.prev = _head;
    }

    public boolean isEmpty() {
        return _head.next == _tail;
    }

    public void add(int index, int value) {
        ListNode current = _head;
        int count = 0;
        while (current != _tail && count < index) {
            current = current.next;
            ++count;
        }

        if (current != _tail && count <= index) {
            ListNode node = new ListNode(value);
            node.next = current.next;
            node.prev = current;
            current.next.prev = node;
            current.next = node;
        }
    }

    public void addHead(int value) {
        ListNode node = new ListNode(value);
        node.next = _head.next;
        node.prev = _head;
        _head.next.prev = node;
        _head.next = node;
    }

    public void addTail(int value) {
        ListNode node = new ListNode(value);
        node.prev = _tail.prev;
        node.next = _tail;
        _tail.prev.next = node;
        _tail.prev = node;
    }

    public void erase(int index) {
        ListNode current = _head;
        int count = 0;
        while (current.next != _tail && count < index) {
            current = current.next;
            ++count;
        }

        if (current.next != _tail && count <= index) {
            current.next.next.prev = current;
            current.next = current.next.next;
        }
    }

    public void eraseHead() {
        if (_head.next != _tail) {
            _head.next.next.prev = _head;
            _head.next = _head.next.next;
        }
    }

    public void eraseTail() {
        if (_head.next != _tail) {
            _tail.prev.prev.next = _tail;
            _tail.prev = _tail.prev.prev;
        }
    }

    public void set(int index, int value) {
        ListNode current = _head.next;
        int count = 0;
        while (current != _tail && count < index) {
            current = current.next;
            ++count;
        }

        if (current != _tail && count <= index) {
            current.value = value;
        }
    }

    public int get(int index) {
        ListNode current = _head.next;
        int count = 0;
        while (current != _tail && count < index) {
            current = current.next;
            ++count;
        }

        if (current != _tail && count <= index) {
            return current.value;
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        _head.next = _tail;
        _tail.prev = _head;
    }

    private class ListNode {
        public int value = 0;
        public ListNode prev = null;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}
